ANTLR/src/Compiler/ContextualAnalysis/TypeCheckerVisitor.java

```diff
@@ -113,24 +114,37 @@ public Integer visitExprBinaryOp(ExprBinaryOpContext ctx) {
 
     @Override
     public Integer visitExprUnaryOp(ExprUnaryOpContext ctx) {
-        Integer exprType = visit(ctx.expr());
-        if (exprType != BOOLTYPE) {
-            String typeName = VOCABULARY.getLiteralName(exprType);
-            String errorMsg = String.format("Incompatible type: Type %s is incompatible on operation NOT.", typeName);
-            errorListener.ThrowError(errorMsg, ctx.op);
-        }
+        checkIfExprIsNumberType(ctx.expr(),
+                BOOLTYPE,
+                ctx.op,
+                typeName -> String.format("Incompatible type: Type %s is incompatible on operation NOT.", typeName));
         return BOOLTYPE;
     }
 
     @Override
     public Integer visitExprMinusPrefix(ExprMinusPrefixContext ctx) {
-        Integer exprType = visit(ctx.expr());
-        if (exprType != NUMTYPE) {
+        checkIfExprIsNumberType(ctx.expr(),
+                NUMTYPE,
+                ctx.op,
+                typeName -> String.format("Incompatible type: Type %s cannot be negative.", typeName));
+        return NUMTYPE;
+    }
+
+    /**
+     * Type checks an expression in an unary operation
+     * @param ctx                 The node in question
+     * @param operator            The unary operator
+     * @param produceErrorMessage The Lambda function should take the string representation of the type of the
+     *                            expression as input, and produce a string, which will be the error message to
+     *                            the user
+     */
+    private void checkIfExprIsNumberType(ExprContext ctx, Integer expectedType, Token operator, Lambda<String, String> produceErrorMessage) {
```

Comment:
Please change the name. I have a suggestion
```suggestion
    private void isExprTypeOfExpectedType(ExprContext ctx, Integer expectedType, Token operator, Lambda<String, String> produceErrorMessage) {
```